src/modes/ModeEstimation.jl

```diff
@@ -0,0 +1,235 @@
+
+Fields:
+
+- `values` is a vector with the resulting point estimates
+- `info_matrix` is the inverse Hessian
+- `optim_result` is the stored Optim.jl results
+- `summary_table` is a summary table with parameters, standard errors, and 
+  t-statistics computed from the information matrix.
+- `lp` is the final likelihood.
+"""
+struct ModeResult{
+    V<:NamedArrays.NamedArray, 
+    M<:Union{Missing, NamedArrays.NamedArray}, 
+    O<:Optim.MultivariateOptimizationResults, 
+    S<:NamedArrays.NamedArray
+}
+    values :: V
+    info_matrix :: M
+    optim_result :: O
+    summary_table :: S
+    lp :: Float64
+end
+
+function Base.show(io::IO, m::ModeResult)
+    show(io, m.summary_table)
+end
+
+"""
+    make_logjoint(model::DynamicPPL.Model, ctx::DynamicPPL.AbstractContext)
+
+Constructs a log density function that accepts a vector `z` and returns 
+a tuple (-likelihood, `varinfo`). The model is run using the provided 
+context `ctx`.
+"""
+function make_logjoint(model::DynamicPPL.Model, ctx::DynamicPPL.AbstractContext)
+    # setup
+    varinfo_init = Turing.VarInfo(model)
+    spl = DynamicPPL.SampleFromPrior()    
+    DynamicPPL.link!(varinfo_init, spl)
+
+    function logπ(z; unlinked = false)
+        varinfo = DynamicPPL.VarInfo(varinfo_init, spl, z)
+
+        unlinked && DynamicPPL.invlink!(varinfo_init, spl)
+        model(varinfo, spl, ctx)
+        unlinked && DynamicPPL.link!(varinfo_init, spl)
+
+        return -DynamicPPL.getlogp(varinfo)
+    end
+
+    return logπ
+end
+
+"""
+    mode_estimation(
+        model::DynamicPPL.Model, 
+        lpf; 
+        optim_options=Optim.Options(),
+        kwargs...
+    )
+
+An internal function that handles the computation of a MLE or MAP estimate.
+
+Arguments: 
+
+- `model` is a `DynamicPPL.Model`.
+- `lpf` is a function returned by `make_logjoint`.
+
+Optional arguments:
+
+- `optim_options` is a `Optim.Options` struct that allows you to change the number
+  of iterations run in an MLE estimate.
+
+"""
+function mode_estimation(
+    model::DynamicPPL.Model, 
+    lpf; 
+    optim_options=Optim.Options(),
+    kwargs...
+)
+    # Do some initialization.
+    b = bijector(model)
+    binv = inv(b)
+
+    spl = DynamicPPL.SampleFromPrior()
+    vi = DynamicPPL.VarInfo(model)
+    init_params = model(vi, spl)
+    init_vals = vi[spl]
+
+    # Construct target function.
+    target(x) = lpf(x)
+    hess_target(x) = lpf(x; unlinked=true)
+
+    # Optimize!
+    M = Optim.optimize(target, init_vals, optim_options)
+
+    # Retrieve the estimated values.
+    vals = binv(M.minimizer)
+
+    # Get the VarInfo at the MLE/MAP point, and run the model to ensure 
+    # correct dimensionality.
+    vi[spl] = vals
+    model(vi) # XXX: Is this a necessary step?
+
+    # Make one transition to get the parameter names.
+    ts = [Turing.Inference.Transition(DynamicPPL.tonamedtuple(vi), DynamicPPL.getlogp(vi))]
+    varnames, _ = Turing.Inference._params_to_array(ts)
```

Comment:
Can't you get the parameter names from `VarInfo` directly? If not, we should change that since it seems weird and inefficient to first convert it to a Transition and then extract the names.

Reply:
Not in a quick way, at least as far as I can tell. If someone's got a good fix other than this (reasonbly good) hack, I'm game.